### ISSUE TITLE
Port CUDA Kernels

### ISSUE THREAD

**Issue by xrsrke** [xrsrke/pipegoose#8]
> Port training CUDA kernels from these librarys, and automatically replace modules in an existing  🤗 `transformers` model with their corresponding CUDA kernel version.
> 

List all CUDA kernels from the sources above and port them (don’t forget to write tests)

- https://github.com/spcl/substation
- https://github.com/NVIDIA/apex
- https://github.com/NVIDIA/Megatron-LM
- https://github.com/NVIDIA/FasterTransformer
- https://github.com/hpcaitech/ColossalAI
- https://github.com/bytedance/lightseq
- https://github.com/bytedance/effective_transformer
- https://github.com/Tencent/TurboTransformers
- https://github.com/microsoft/fastseq/tree/main/fastseq/clib/cuda
- https://github.com/microsoft/DeepSpeed/tree/master/csrc

Jit

- https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/jit.py

**Comment by isamu-isozaki** [comment]
Hi! So a question about how you want the cuda kernels integrated in. In substation, they do optimizations by "generating" cuda files specific to the dimensions of each kernel(from what I understand from looking at https://github.com/spcl/substation/blob/master/pytorch_module/test_softmax.py) so basically it makes cuda code specific for each function so it's faster but may be messier.

Bitsandbytes does it but just loading the .so files from a given location and deepspeed does it by I think building the cstr when doing pip install. So a bit slower but more general.

There might be more ways to do it but which way do you think will work the best for you?

**Comment by isamu-isozaki** [comment]
Just checked colossalai, seems like they have functions called op_builders that they use to build certain cuda libraries.

**Comment by xrsrke** [comment]
@isamu-isozaki I think we should do something like this. What are the pros and cons of substation? What do you think we should use? If there are some operations that `substation` is really good at, we could do both `substation` and manually port kernels.

```
class KernelBuilder:
    def load(self):
        pass


class FusedOp(KernelBuilder):
    def absolute_name(self):
         # NOTE: the absolute path to the kernel
         pass

fused_op = FusedOp().load()
outputs = fused_op(inputs)
```

**Comment by isamu-isozaki** [comment]
@xrsrke I think substation's method in general is faster, but it needs you to generate a new cuda file for each possible shape of tensor. So the main disadvantage is it's not clean I think(My guess is just changing batch size will need a new cuda script if we were to just copy).

I think the way you are doing is similar to colossalai's and deepspeed's ver which we can definitely do. I do remember that setting up colossalai is pretty troublesome compared to say deepspeed which I'm not sure why but we can probably cross that bridge when we get there. I think this approach is more general but might be slightly slower than say substation. I think we can probably start with this approach and if we want, extend to substation and build kernels specific to a certain dimension input

**Comment by isamu-isozaki** [comment]
do you think this makes sense?

**Comment by xrsrke** [comment]
@isamu-isozaki Could you try to benchmark the two approaches? Try fusing a softmax using substation, then compare the two approaches... (also, it could be that some operations are performed better by substitution, while others are more efficient when written manually. We should take this into account while benchmarking)

Also, I've just added GPT-NeoX's kernel to the issue mentioned above.
